common/logging.hpp

```diff
@@ -0,0 +1,159 @@
+#pragma once
+
+#include <atomic>
+#include <unordered_map>
+#include <vector>
+#include <chrono>
+#include <memory>
+#include "phonebook.hpp"
+#include "cpu_timer.hpp"
+#include "record_types.hpp"
+
+namespace ILLIXR {
+
+	/**
+	 * @brief The ILLIXR logging service.
+	 *
+	 * This has two advantages over printf logging. It has lower
+	 * overhead (because it goes into a database), won't result in
+	 * spliced messages (no stdout race-conditions), and is used
+	 * uniformly by ILLIXR components.
+	 */
+	class c_metric_logger : public phonebook::service {
+	public:
+		/**
+		 * @brief Writes one log record.
+		 *
+		 * See the default record types record_types.hpp. Note that this is an 'open system'; new
+		 * log record types can be added in a component according to the instructions in
+		 * record_types.hpp.
+		 */
+		template <typename Record>
+		void log(std::unique_ptr<const Record>&& r) {
+			log2(&Record::type_descr, std::unique_ptr<const record>(std::move(r)));
+		}
+
+		/**
+		 * @brief Writes many of the same type of log record.
+		 *
+		 * This is more efficient than calling log many times.
+		 */
+		template <typename Record>
+		void log_many(std::vector<std::unique_ptr<const Record>>&& rs) {
+			std::vector<std::unique_ptr<const record>> rs2;
+			for (std::unique_ptr<const Record>& r : rs) {
+				rs2.emplace_back(std::move(r));
+			}
+			// I think (hope) the compiler will optimize this copy-and-cast away.
+			// Unfortunately, there is no way to cast every element of a vector all at once according to
+			// https://stackoverflow.com/a/20310090/1078199
+			log_many2(&Record::type_descr, std::move(rs2));
```

Comment:
Where does this call end up?

Reply:
`c_metric_logger` is an abstract class, so there will be some class that extends that interface (call it `impl_class`).
- `c_metric_logger::log` calls `impl_class::log2`
- `c_metric_logger::log_many` calls `impl_class::log_many2`